include/hipSYCL/runtime/instrumentation.hpp

```diff
@@ -0,0 +1,153 @@
+  }
+};
+
+class instrumentation {
+public:
+  virtual ~instrumentation() = default;
+};
+
+template<typename T>
+inline constexpr bool is_instrumentation_v
+    = std::is_convertible_v<std::remove_cv_t<T> *, instrumentation *>;
+
+class not_instrumented: public std::exception {
+public:
+  explicit not_instrumented(const std::type_info &instr_type): _instr_type(&instr_type) {}
+  const char *what() const noexcept override { return _instr_type->name(); }
+  const std::type_info &instrumentation() const { return *_instr_type; }
+
+private:
+  const std::type_info *_instr_type;
+};
+
+// Provides synchronization points for instrumentations between front- and backend threads.
+//
+// Thread safety: Before submission - single-threaded access only
+//                After submission: - concurrent access, but do not call instrument<>()
+class instrumentation_set {
+public:
+  // Create a producer-consumer channel for a particular instrumentation
+  template<typename Instr>
+  void instrument() {
+    static_assert(is_instrumentation_v<Instr>);
+    if (_instrs.find(typeid(Instr)) == _instrs.end()) {
+      _instrs.emplace(typeid(Instr), future_instrumentation{});
+    }
+  }
+
+  // Whether instrument<Instr>() has been called before.
+  template<typename Instr>
+  bool is_instrumented() const {
+    static_assert(is_instrumentation_v<Instr>);
+    return _instrs.find(typeid(Instr)) != _instrs.end();
+  }
+
+  // Resolve the instrument<Instr>() promise by providing an instrumentation instance.
+  // instrument<Instr>() must have been called before submission.
+  template<typename Instr, typename T>
+  void provide(std::unique_ptr<T> instr) {
+    static_assert(is_instrumentation_v<Instr>);
+    static_assert(std::is_convertible_v<T*, Instr*>);
+    if (auto iter = _instrs.find(typeid(Instr)); iter != _instrs.end()) {
+      iter->second.provide(std::move(instr));
+    } else {
+      throw not_instrumented(typeid(Instr));
+    }
+  }
+
+  // Wait until another thread provide()s an instance of Instr.
+  // instrument<Instr>() must have been called before submission.
+  template<typename Instr> const Instr &await() const {
+    static_assert(is_instrumentation_v<Instr>);
+    if (auto iter = _instrs.find(typeid(Instr)); iter != _instrs.end()) {
+      // static_cast: type is statically checked in resolve()
+      return static_cast<const Instr &>(iter->second.await());
+    } else {
+      throw not_instrumented(typeid(Instr));
+    }
+  }
+
+private:
+  class future_instrumentation {
+  public:
+    future_instrumentation(): _future{_promise.get_future()} {}
+    void provide(std::unique_ptr<instrumentation> instr) { _promise.set_value(std::move(instr)); }
+    const instrumentation &await() const { return *_future.get(); }
+
+  private:
+    std::promise<std::unique_ptr<instrumentation>> _promise;
+    std::shared_future<std::unique_ptr<instrumentation>> _future;
+  };
```

Comment:
There are two distinct wait states:
1. The time between creation of the `dag_node` in the queue's thread, where the `sycl::event` first becomes visible to the user and the instant where the `inorder_queue` instantiates the backend-specific instrumentation type.
2. The time from instantiation of the instrumentation to arrival of a specific event, e.g. `command_finished`. This is backend-specific; for CUDA, it blocks on `cudaEventSynchronize()`.

What you see here is the implementation of state 1, were the backend is not involved yet (and probably unknown). Since a user can call `event::get_profiling_info()` immediately after submission, they need something to block on.